### PR TITLE
Fix TF-TRT control edges between TRTEngineOp nodes

### LINKED ISSUE #181: Segmentation fault when converting TF model to TF-TRT in TF2.1
## Env
* TensorRT docker: https://docs.nvidia.com/deeplearning/sdk/tensorrt-container-release-notes/rel_19-10.html#rel_19-10
* TF2.1
* GPU: RTX2080

## Code

```
conversion_params = trt.DEFAULT_TRT_CONVERSION_PARAMS._replace(precision_mode=trt.TrtPrecisionMode.FP32, max_workspace_size_bytes=4000000000)                                                           
converter = trt.TrtGraphConverterV2(input_saved_model_dir=model_dir,                                                                                                                                    
                                        conversion_params=conversion_params)                                                                                                                                
converter.convert()                                                                                                                                                                                     
converter.save(output_saved_model_dir=FLAGS.tftrt_model_dir)             
```

## Error message

```

....Converting to TF-TRT FP32...
2020-02-22 06:24:54.194963: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libnvinfer.so.6
INFO:tensorflow:Linked TensorRT version: (6, 0, 1)
I0222 06:24:54.195068 140612687255360 trt_convert.py:200] Linked TensorRT version: (6, 0, 1)
INFO:tensorflow:Loaded TensorRT version: (6, 0, 1)
I0222 06:24:54.195302 140612687255360 trt_convert.py:201] Loaded TensorRT version: (6, 0, 1)
2020-02-22 06:24:57.597559: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-02-22 06:24:57.597931: I tensorflow/core/grappler/devices.cc:55] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 1
2020-02-22 06:24:57.598014: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2020-02-22 06:24:57.598443: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-02-22 06:24:57.598817: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce RTX 2080 with Max-Q Design computeCapability: 7.5
coreClock: 1.095GHz coreCount: 46 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 357.69GiB/s
2020-02-22 06:24:57.598843: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2020-02-22 06:24:57.598854: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-02-22 06:24:57.598897: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2020-02-22 06:24:57.598910: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2020-02-22 06:24:57.598926: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2020-02-22 06:24:57.598939: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2020-02-22 06:24:57.598950: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
....
2020-02-22 06:24:59.442725: I tensorflow/compiler/tf2tensorrt/segment/segment.cc:460] There are 416 ops of 8 different types in the graph that are not converted to TensorRT: BatchToSpaceND, SpaceToBatchND, Placeholder, NoOp, Switch, Identity, Merge, StridedSlice, (For more information see https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html#supported-ops).
2020-02-22 06:24:59.590980: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:636] Number of TensorRT candidate segments: 77
2020-02-22 06:24:59.623082: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 0 consisting of 35 nodes by StatefulPartitionedCall/TRTEngineOp_0.
2020-02-22 06:24:59.623329: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 1 consisting of 57 nodes by StatefulPartitionedCall/TRTEngineOp_1.
2020-02-22 06:24:59.623489: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 2 consisting of 9 nodes by StatefulPartitionedCall/single_stage_model0/resnet_block1/conv1d_3/TRTEngineOp_2.
2020-02-22 06:24:59.623567: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 3 consisting of 14 nodes by StatefulPartitionedCall/single_stage_model0/resnet_block1/TRTEngineOp_3.
2020-02-22 06:24:59.623652: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 4 consisting of 9 nodes by StatefulPartitionedCall/single_stage_model0/resnet_block2/conv1d_5/TRTEngineOp_4.

.....



2020-02-22 06:09:26.968512: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 58 consisting of 35 nodes by TRTEngineOp_58.
2020-02-22 06:09:26.968593: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:737] Replaced segment 59 consisting of 9 nodes by StatefulPartitionedCall/single_stage_model3/resnet_block1/conv1d_69/TRTEngineOp_59.
Fatal Python error: Segmentation fault

Current thread 0x00007f0f27e09740 (most recent call first):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/grappler/tf_optimizer.py", line 59 in OptimizeGraph
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/compiler/tensorrt/trt_convert.py", line 935 in _run_conversion
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/compiler/tensorrt/trt_convert.py", line 992 in convert
  File "/app/models/ms_tcn/tensorflow/inference.py", line 192 in convert_to_tftrt
  File "/app/models/ms_tcn/tensorflow/inference.py", line 198 in run_inference
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250 in _run_main
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299 in run
  File "/app/models/ms_tcn/tensorflow/inference.py", line 332 in <module>
  File "/usr/lib/python3.6/runpy.py", line 85 in _run_code
  File "/usr/lib/python3.6/runpy.py", line 193 in _run_module_as_main
Segmentation fault (core dumped)
```

## More log with gdb

```
Thread 1 "inference.py" received signal SIGSEGV, Segmentation fault.
0x00007fff921bb560 in tensorflow::Node::name() const () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/../libtensorflow_framework.so.2
(gdb) where
#0  0x00007fff921bb560 in tensorflow::Node::name() const () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/../libtensorflow_framework.so.2
#1  0x00007fff9990cf66 in tensorflow::tensorrt::convert::UpdateToEngineNode(std::vector<tensorflow::tensorrt::convert::EngineInfo, std::allocator<tensorflow::tensorrt::convert::EngineInfo> > const&, unsigned long, std::vector<tensorflow::Node*, std::allocator<tensorflow::Node*> > const&, bool, std::string const&, tensorflow::Node**, int*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#2  0x00007fff9990efc5 in tensorflow::tensorrt::convert::CreateTRTNode(tensorflow::tensorrt::convert::ConversionParams const&, std::vector<tensorflow::tensorrt::convert::EngineInfo, std::allocator<tensorflow::tensorrt::convert::EngineInfo> > const&, int, int, tensorflow::Graph*, nvinfer1::IGpuAllocator*, std::vector<tensorflow::Node*, std::allocator<tensorflow::Node*> >*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#3  0x00007fff99914041 in tensorflow::tensorrt::convert::ConvertAfterShapes(tensorflow::tensorrt::convert::ConversionParams const&) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#4  0x00007fff9994e34b in tensorflow::tensorrt::convert::TRTOptimizationPass::Optimize(tensorflow::grappler::Cluster*, tensorflow::grappler::GrapplerItem const&, tensorflow::GraphDef*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#5  0x00007fff9c2c1937 in tensorflow::grappler::MetaOptimizer::RunOptimizer(tensorflow::grappler::GraphOptimizer*, tensorflow::grappler::Cluster*, tensorflow::grappler::GrapplerItem*, tensorflow::GraphDef*, tensorflow::grappler::MetaOptimizer::GraphOptimizationResult*) () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#6  0x00007fff9c2c2ccd in tensorflow::grappler::MetaOptimizer::OptimizeGraph(tensorflow::grappler::Cluster*, tensorflow::grappler::GrapplerItem const&, tensorflow::GraphDef*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#7  0x00007fff9c2c4714 in tensorflow::grappler::MetaOptimizer::Optimize(tensorflow::grappler::Cluster*, tensorflow::grappler::GrapplerItem const&, tensorflow::GraphDef*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#8  0x00007fff956bec07 in TF_OptimizeGraph(GCluster, tensorflow::ConfigProto const&, tensorflow::MetaGraphDef const&, bool, std::string const&, bool, TF_Status*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#9  0x00007fff956c37d6 in _wrap_TF_OptimizeGraph () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#10 0x00000000005097cf in ?? ()
#11 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#12 0x0000000000507125 in ?? ()
#13 0x0000000000508fa0 in ?? ()
#14 0x000000000050999d in ?? ()
#15 0x000000000050c36e in _PyEval_EvalFrameDefault ()
#16 0x0000000000508c69 in ?? ()
#17 0x000000000050999d in ?? ()
#18 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#19 0x0000000000507125 in ?? ()
#20 0x0000000000508fa0 in ?? ()
#21 0x000000000050999d in ?? ()
#22 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#23 0x0000000000508c69 in ?? ()
#24 0x000000000050999d in ?? ()
#25 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#26 0x0000000000507125 in ?? ()
#27 0x0000000000508fa0 in ?? ()
#28 0x000000000050999d in ?? ()
#29 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#30 0x0000000000508c69 in ?? ()
#31 0x000000000050999d in ?? ()
#32 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#33 0x0000000000507125 in ?? ()
#34 0x0000000000508fa0 in ?? ()
#35 0x000000000050999d in ?? ()
#36 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#37 0x0000000000507125 in ?? ()
#38 0x0000000000515904 in ?? ()
#39 0x00000000005097cf in ?? ()
#40 0x000000000050b4a9 in _PyEval_EvalFrameDefault ()
#41 0x0000000000507125 in ?? ()
---Type <return> to continue, or q <return> to quit---

```

### PR DESCRIPTION
**Pull request by tfeher**
This PR fixes https://github.com/tensorflow/tensorrt/issues/181. 

There is a bug in handling control edges between two TRTEngineOp nodes. Tagging @aaroey for review since he was the last one changing the [input edge handling](https://github.com/tensorflow/tensorflow/commit/93bf7372515057204fdadc4db367cfb7136f8e85), and this PR is about doing the same for the output edges. 

**Problem description**
After the graph is segmented, an EngineInfo struct is created for each segment, which list the connections with nodes outside of the segment.

1. For some reason input control edges from const nodes are omitted here: https://github.com/tensorflow/tensorflow/blob/c1915ebc2ad8115636f88b69884194da2f111fa8/tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc#L174-L180
2. At the same time output control edges from const nodes are not omitted: https://github.com/tensorflow/tensorflow/blob/c1915ebc2ad8115636f88b69884194da2f111fa8/tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc#L222-L227

This leads fatal error in the following case:
- there is a control edge from TRTEngineOp_0 --> TRTEngineOp_1 and
- the original node int TRTEngineOp_0 was const.

So let's consider such a case:
- const node A in TRTEgineOp_0
- node B in TRTEngineOp_1, and 
- a control edge A->B in between. 

The error occurs the following way: 
Because of (2.), the A->B edge is listed for TRTEngineOp_0, and we try to rewire this edge for TRTEngineOp_0 [here](https://github.com/tensorflow/tensorflow/blob/c1915ebc2ad8115636f88b69884194da2f111fa8/tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc#L515). `UpdateToEngineNode` tries to look up edge A->B in the EnigineInfo->connections list for TRTEngineOp_1. This was not not added to the list (1.), and UpdateToEngineNode returns with a fatal error.

**Fix**
The problem can be fixed by omitting control edges from const nodes. Alternatively one could consider keeping the control input edges from const nodes. 

**Note**
It seems the way the control edges are wired for the same high level (Keras) network changes between different TensorFlow versions. It can also change during Grappler transformations. The TensorFlow version that is used for NVIDIA's NGC 20.02 docker image happens to wire problematic control edges, and therefore that causes TF-TRT to fail. 

The current upstream master version has changed control edge handling, and the problem does not show up if you compile that.

**Code to reproduce teh issue**
To reproduce the problem, run the following code in the nvcr.io/nvidia/tensorflow:20.02-tf2-py3 docker container (or see https://github.com/tensorflow/tensorrt/issues/181). 

```python
import tensorflow as tf
from tensorflow.python.compiler.tensorrt import trt_convert as trt

inputs = tf.keras.Input(shape=(512, 512, 1))
x = tf.keras.layers.Conv2D(filters=64, kernel_size=(3, 3),
                           activation=None, padding='same', use_bias=False)(inputs)
output = tf.keras.layers.Conv2DTranspose(filters=32, kernel_size=(3, 3),
                                         strides=(2, 2), padding='same',
                                         activation=tf.nn.relu)(x)
model = tf.keras.Model(inputs=inputs, outputs=output)

model.save_weights("results/checkpoint")
tf.keras.models.save_model(model, "results/SavedModel", save_format="tf",
                           overwrite=True, include_optimizer=False)
print("Model saved to SavedModel")

converter = trt.TrtGraphConverterV2(input_saved_model_dir="results/SavedModel")
converter.convert()
converter.save("results/trt_model")
print("Model exported to TRT")
```

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc [hunk 1]
```diff
@@ -221,9 +221,11 @@ Status GetEngineInfo(const Graph* g,
       }
       if (edge->IsControlEdge()) {
         // Control output.
-        info->connections.emplace_back(output_node->name(), output_node->id(),
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: Shall we have a helper function, say ShallKeepConrolEdgeFrom(Node* input_node), for here and the place that handle node? 
According to the change for the input node, it seems that the purpose of removing such edges  is to allow removing unused constants. But how does this expose unused constants is something I don't understand, still trying to find out about this.

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: I have added the ShallKeepControlEdgeFrom helper function.

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc [hunk 2]
```diff
@@ -103,6 +103,15 @@ std::pair<TfGpuId, PlatformGpuId> GetFirstValidDeviceId() {
   return std::make_pair(TfGpuId(-1), PlatformGpuId(-1));
 }
 
+// Returns false for const nodes (we intend to drop control edges from those).
+bool ShallKeepControlEdgeFrom(const Node* input_node) {
+  if (!input_node) {
+    VLOG(2) << "Node pointer is null, this should not happen";
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: Shouldn't we use LOG(FATAL) instead then?

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: Yes, we could.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**